### ISSUE TITLE
Unable to know status of the event

### ISSUE THREAD

**Issue by JulienChavee** [getsentry/sentry-php#886]
I'm trying to migrate from V1 to V2.

In V1 of this SDK, we were able to know if an exception failed to be sent via some methods or by checking if eventID is null.

In V2, it seems impossible to know if the exception was correctly sent to Sentry.

```
$eventID = \Sentry\captureException($myException);
```
This code will always give me an EventID even if the exception wasn't correctly sent.

I checked a bit in the code and I saw that in Transport/HttpTransport::send() you are doing this :

```
 $promise->wait(false);
```
Passing false to this function means that you don't want to check the status the request : "bool $unwrap Whether to return resolved value / throw reason or not"

If I change this value to "true", I can now catch error when the exception wasn't correctly sent (as the method is throwing an Exception).

Now, I would like to know if it is possible (out of the box) to check if the exception is correctly sent? 
And furthermore, is it intended to not warn the user that the exception isn't correctly sent?

**Comment by ste93cry** [comment]
>This code will always give me an EventID even if the exception wasn't correctly sent.

Even though I totally agree that it makes no sense, Unified API works that way and so this behavior is expected

>Now, I would like to know if it is possible (out of the box) to check if the exception is correctly sent?

Out-of-the-box it's not possible at the moment of writing to know if an exception was sent correctly or not, but you can build your own transport that instead of silently catching all errors will do whatever you think is best. See the next part of my reply for what you can expect in the next version to solve or at least improve the handling of this issue

>And furthermore, is it intended to not warn the user that the exception isn't correctly sent?

Complex topic here. We catch any exception that occurs during the sending of the event to avoid breaking the application, but it makes sense to log it somehow, be it with a simple logger or by capturing it and sending it as any other error. What I plan to do in `2.3` is to support a PSR-3 logger in some of the critical parts of the library so that we can log what is happening. Would that be enough?

**Comment by Jean85** [comment]
In general, raising an exception while reporting an error to Sentry is a risky default behavior: it may happen that the error is a network failure, and that would create a cascade of exceptions, since even Sentry server wouldn't be reachable. 

IMHO the current behavior, as a "best effort", is a good one. Both approaches proposed by @ste93cry (custom transport and logger) are good :+1:

**Comment by JulienChavee** [comment]
Many thanks for the answers.

We will probably write our custom transport (it the next idea if there is no solution out-of-the-box).

I understand that it can be a risky behaviour to raise an exception during the process as it could lead to a cascade of exceptions. On our side we should not have this problem we have our own Exceptions catching system as we "manually" send exceptions to Sentry (and we can catch exceptions during this process to log them somewhere else than in Sentry [and send the "original" exception to another place too]).

To finish, a logger is a great idea 👍

**Comment by helhum** [comment]
> Even though I totally agree that it makes no sense, Unified API works that way and so this behavior is expected

Sorry to kick in here as well. But the [current state of the code](https://github.com/getsentry/sentry-php/blob/be044bfa83ff98c5dd0ea0a7ff243ff0eade721c/src/Transport/HttpTransport.php#L112) is returning `null` when an error happens. Is this a bug then?

> What I plan to do in 2.3 is to support a PSR-3 logger in some of the critical parts of the library so that we can log what is happening. Would that be enough?

If the current behavior is changed again to always return an eventId, then a PSR-3 logger would be great. I'll try to prepare a PR for this.

**Comment by ste93cry** [comment]
>Sorry to kick in here as well. But the current state of the code is returning null when an error happens. Is this a bug then?

It's a regression indeed that got fixed in #905
